pkg/query-service/app/opamp/pipeline_builder.go

```diff
@@ -0,0 +1,196 @@
+	},
+	1: {
+		Name:    "batch",
+		Enabled: true,
+	},
+}
+
+func updatePipelineSpec(signal string, name string, enabled bool) {
+	switch signal {
+	case "metrics":
+		lockMetricsPipelineSpec.Lock()
+		defer lockMetricsPipelineSpec.Unlock()
+
+		for i := 0; i < len(metricsPipelineSpec); i++ {
+			p := metricsPipelineSpec[i]
+			if p.Name == name {
+				p.Enabled = enabled
+				metricsPipelineSpec[i] = p
+			}
+		}
+	case "traces":
+		lockTracesPipelineSpec.Lock()
+		defer lockTracesPipelineSpec.Unlock()
+
+		for i := 0; i < len(tracesPipelineSpec); i++ {
+			p := tracesPipelineSpec[i]
+			if p.Name == name {
+				p.Enabled = enabled
+				tracesPipelineSpec[i] = p
+			}
+		}
+	default:
+		return
+	}
+
+}
+
+// AddToTracePipeline to enable processor in traces pipeline
+func AddToTracePipelineSpec(processor string) {
+	updatePipelineSpec("traces", processor, true)
+}
+
+// RemoveFromTracePipeline to remove processor from traces pipeline
+func RemoveFromTracePipelineSpec(name string) {
+	updatePipelineSpec("traces", name, false)
+}
+
+// AddToMetricsPipeline to enable processor in traces pipeline
+func AddToMetricsPipelineSpec(processor string) {
+	updatePipelineSpec("metrics", processor, true)
+}
+
+// RemoveFromMetricsPipeline to remove processor from traces pipeline
+func RemoveFromMetricsPipelineSpec(name string) {
+	updatePipelineSpec("metrics", name, false)
+}
+
+func checkDuplicates(pipeline []interface{}) bool {
+	exists := make(map[string]bool, len(pipeline))
+	zap.S().Debugf("checking duplicate processors in the pipeline:", pipeline)
+	for _, processor := range pipeline {
+		name := processor.(string)
+		if _, ok := exists[name]; ok {
+			return true
+		}
+
+		exists[name] = true
+	}
+	return false
+}
+
+func buildPipeline(signal Signal, current []interface{}) ([]interface{}, error) {
+	var spec map[int]pipelineStatus
+
+	switch signal {
+	case Metrics:
+		spec = metricsPipelineSpec
+		lockMetricsPipelineSpec.Lock()
+		defer lockMetricsPipelineSpec.Unlock()
+	case Traces:
+		spec = tracesPipelineSpec
+		lockTracesPipelineSpec.Lock()
+		defer lockTracesPipelineSpec.Unlock()
+	default:
+		return nil, fmt.Errorf("invalid signal")
+	}
+
+	pipeline := current
+	// create a reverse map of existing config processors and their position
+	existing := map[string]int{}
+	for i, p := range current {
+		name := p.(string)
+		existing[name] = i
+	}
+
+	// create mapping from our tracesPipelinePlan (processors managed by us) to position in existing processors (from current config)
+	// this means, if "batch" holds position 3 in the current effective config, and 2 in our config, the map will be [2]: 3
+	specVsExistingMap := map[int]int{}
+
+	// go through plan and map its elements to current positions in effective config
+	for i, m := range spec {
+		if loc, ok := existing[m.Name]; ok {
+			specVsExistingMap[i] = loc
+		}
+	}
+
+	lastMatched := -1
+	inserts := 0
+
+	// go through plan again in the increasing order
+	for i := 0; i < len(spec); i++ {
```

Comment:
I am trying to understand this logic. Let's say I have ['spanmetrics', 'attribute', 'batch'] in my original config. And now, the user wants to enable the sampling, IIUC, this adds the 'signoz_tail_sampling' in the first position, which will affect the span metrics. How is this situation handled generally in this implementation?

Reply:
The overrides can happen at 2 places.  In our config (will get to this later) and directly in config file. 

In this case,  user can add `signoz_tail_sampling` manually in the list to enforce an order.   If we find signoz_tail_sampling we would not add it again or change its position. 

`processor: [spanmetrics, signoz_tail_sampling]  `

If config file has the above order, it will be maintained.

This does require some documentation for user but addresses the problem of coordination.

Reply:
Regarding some of the dependencies that we wish to enforce, we can also use tracePipelinePlan{} to ensure a correct order.  All the processors in this map need not be custom - for e.g. signoz_tail_sampling.

Reply:
> user can add signoz_tail_sampling manually in the list to enforce an order

Why does user need to intervene manually? For instance, they didn't have the `signoz_tail_sampling` earlier and they are configuring it from UI. In this case it adds in order which results in incorrect span metrics.

>Regarding some of the dependencies that we wish to enforce, we can also use tracePipelinePlan{} to ensure a correct order.

Take the above example and what does this look like?

Reply:
> Why does user need to intervene manually? For instance, they didn't have the signoz_tail_sampling earlier and they are configuring it from UI. In this case it adds in order which results in incorrect span metrics.

Since the config editing is shared between UI (signoz) and the user (directly in the file). some level of co-ordination would be required. What we have implemented so far, covers basic use case where we don't overwrite what is already there.  

The current assumption is user would not have dependencies on our order and we don't on user edits.  But if there is such a scenario, we would have to co-ordinate.   IMO, as long as we can explain how we edit config files, users should be able to make their edits.  

In case you have suggestions for better way to handle co-ordinated edits with user, please create a separate issue if thats ok.

Reply:
The discussion has gone in a different direction. Let me bring up my original point; this logic adds a tail sampling processor before the spanmetrics and spanmetrics is the most basic use case. Adding a tail sampling processor before span metrics is incorrect. How do we address it? I don't think the answer should be we communicate to users and expect it to resolve by them.

Reply:
@mindhash will create a new issue to track this.